File: admin/ui/screens/collections.js
// Collections: the four categories the shop is built around, and its
// all-products page. The list shows each one's name, intro, product count
// and circle picture. One collection opens with its words, its homepage
// circle and shelf, and its products in the shop's order. A save writes Site
// text, Homepage and Navigation together (PUT collections/<key>); moving a
// product saves at once through products/reorder. The categories themselves
// are code, so none can be added here.
import { api } from "../lib/api.js";
import { h, clear, thumb, useCss } from "../lib/dom.js";
import { mountEditor } from "../lib/editor.js";
import { field, iconBtn } from "../lib/forms.js";
import { banner, toast, guard } from "../lib/ui.js";
import { icon } from "../icons.js";

const NO_ADD = "Categories are wired into the shop's code: filters, card lines and breadcrumbs.";

// The collection keys the server's schema knows, in its order: one set of
// Site text fields per key, so nothing here lists them by hand.
function keys(app) {
  return app.state.schema.resources.copy.fields
    .map((f) => (f.path.match(/^\/categories\/([^/]+)\/label$/) || [])[1]).filter(Boolean);
}

const plural = (n, one, many) => n + " " + (n === 1 ? one : many);

export function render(main, ctx) {
  useCss("collections");
  if (!ctx.sub) return list(main, ctx);
  if (!keys(ctx.app).includes(ctx.sub)) return missing(main, ctx.sub);
  return detail(main, ctx);
}

function missing(main, key) {
  guard.clear();
  clear(main).append(
    h("div", { class: "page-head" }, h("div", {},
      h("a", { class: "back", href: "#/collections" }, "Collections"),
      h("h1", { tabindex: "-1" }, "Collection not found"))),
    banner({ tone: "critical", title: "There is no collection called " + key + ".",
      items: ["The shop's collections are its four categories and the all-products page."] }));
}

async function list(main) {
  guard.clear();
  const msgs = h("div", { class: "msgs" });
  const grid = h("div", { class: "coll-grid" });
  clear(main).append(
    h("div", { class: "page-head" },
      h("div", {}, h("h1", { tabindex: "-1" }, "Collections"),
        h("p", { class: "sub" }, "The shop's categories and its all-products page: their names and intros, their homepage circles and shelves, and the order of their products.")),
      h("div", { class: "head-actions coll-add" },
        h("button", { class: "btn", type: "button", disabled: true, "aria-describedby": "coll-add-why" }, icon("plus", 16), "Add collection"),
        h("p", { class: "help", id: "coll-add-why" }, NO_ADD))),
    msgs, grid);
  let res;
  try {
    res = await api("GET", "collections");
  } catch (e) {
    msgs.append(banner({ tone: "critical", title: "The collections did not load.", items: [e.message] }));
    return;
  }
  grid.append(...res.items.map(card));
}

function card(it) {
  const d = it.data;
  const circle = d.circle || {};
  const facts = [plural(it.published, "product", "products") + " on the site"];
  if (it.drafts) facts.push(plural(it.drafts, "draft", "drafts"));
  if (it.shelf_shows) facts.push("homepage shelf of " + plural(it.shelf_shows.cards, "card", "cards"));
  return h("section", { class: "card coll-card" },
    h("div", { class: "coll-pic" + (circle.cutout ? " cut" : "") },
      circle.image ? h("img", { src: "/" + circle.image, alt: "", width: "72", height: "72", loading: "lazy" })
        : icon(it.key === "all" ? "grid" : "image", 22)),
    h("div", { class: "coll-body" },
      h("h2", {}, h("a", { href: "#/collections/" + it.key }, d.label || it.name)),
      it.key === "all" ? h("p", { class: "coll-kind" }, "All-products page") : null,
      d.intro ? h("p", { class: "coll-intro" }, d.intro) : h("p", { class: "coll-intro muted" }, "No intro yet."),
      h("p", { class: "coll-meta" }, facts.join(", ") + ".")));
}

function priceText(d) {
  if (Array.isArray(d.sizes) && d.sizes.length > 1) {
    const ps = d.sizes.map((s) => s.price);
    return "AED " + Math.min(...ps) + " to " + Math.max(...ps);
  }
  return "AED " + d.price;
}

const ARROW = String.fromCharCode(0x2192);

function detail(main, { sub: key, app }) {
  let meta = null;
  const prodCard = h("section", { class: "card" });
  mountEditor(main, app, {
    back: { href: "#/collections", label: "Collections" },
    heading: (d) => (d && d.label) || key,
    subtitle: "Saving updates Site text, Homepage and Navigation together and rebuilds the site.",
    load: async () => {
      meta = await api("GET", "collections/" + key);
      return { rev: meta.rev, data: meta.data, meta };
    },
    put: (data, rev) => api("PUT", "collections/" + key, { body: { data }, rev }),
    view: () => "/" + meta.href,
    actions: () => [h("a", { class: "btn", href: "/" + meta.href, target: "_blank", rel: "noopener noreferrer" }, icon("external", 16), "View on store")],
    form: (ctx, data, m) => {
      ctx.root = data;
      const defs = new Map(m.fields.map((f) => [f.path, f]));
      const put = (p) => (defs.has(p) ? field(defs.get(p), data, p, ctx) : null);
      const shelfNote = h("p", { class: "coll-preview" });
      // What the homepage will draw, worked out as build.py does: the first
      // cards up to the limit, and the link with {n} counted.
      const refresh = () => {
        const s = data.shelf;
        if (!s) return;
        const n = m.published;
        const cards = Number.isInteger(s.limit) ? Math.min(s.limit, n) : n;
        const link = String(s.link_label || "").split("{n}").join(String(n));
        shelfNote.textContent = "The homepage shows " + plural(cards, "card", "cards") + " of the " + n +
          " published, then a link that reads " + link + " " + ARROW + ".";
      };
      const none = (what) => h("p", { class: "muted" }, key === "all" ? "The all-products page has no " + what + " of its own." : "No " + what + " on the homepage belongs to this collection.");
      const words = h("section", { class: "card" }, h("h2", {}, "Name and intro"), put("/label"), put("/crumb"), put("/intro"));
      const shelf = h("section", { class: "card" }, h("h2", {}, "Homepage shelf"),
        data.shelf ? [put("/shelf/heading"), put("/shelf/limit"), put("/shelf/link_label"), shelfNote] : none("shelf"));
      const circle = h("section", { class: "card" }, h("h2", {}, "Homepage circle"),
        data.circle ? [h("p", { class: "sub coll-sub" }, "The circle under the homepage banner and the link in the category bar."),
          put("/circle/label"), put("/circle/tint"), put("/circle/cutout"), put("/circle/image")] : none("circle"));
      const wrap = h("div", {}, h("div", { class: "two-col" }, h("div", {}, words, shelf), h("div", {}, circle)), prodCard);
      wrap.addEventListener("input", refresh);
      wrap.addEventListener("change", refresh);
      refresh();
      return wrap;
    },
  });
  productsCard(prodCard, key);
}

// The collection's products in the shop's order, each moved up or down with
// a save at once (products/reorder, guarded by the product list's rev).
// Every category keeps its own order, so the all-products page points there.
function productsCard(box, key) {
  if (key === "all") {
    clear(box).append(h("h2", {}, "Products"),
      h("p", { class: "coll-note" }, "Every published product is on this page. Each category keeps its own order: open a category to change it."));
    return;
  }
  let data = null;
  let focusAfter = null;
  let moving = false;
  const msgs = h("div", { class: "msgs" });
  const table = h("div", { class: "table-wrap" });
  clear(box).append(h("h2", {}, "Products"),
    h("p", { class: "sub coll-sub" }, "In the shop's order, which the collection page and the homepage shelf follow. A move saves at once and rebuilds the site."),
    msgs, table);

  const load = async () => {
    data = await api("GET", "products?category=" + encodeURIComponent(key));
    draw();
  };
  const moveBtn = (items, i, dir) => {
    const { id, data: d } = items[i];
    const b = iconBtn(dir < 0 ? "up" : "down", "Move " + (d.name || id) + (dir < 0 ? " up" : " down"),
      dir < 0 ? i === 0 : i === items.length - 1, () => move(i, dir));
    b.dataset.move = id + (dir < 0 ? ":up" : ":down");
    return b;
  };
  function draw() {
    clear(table);
    const items = data.items;
    if (!items.length) { table.append(h("p", { class: "empty" }, "No products in this collection yet.")); return; }
    table.append(h("table", { class: "table" },
      h("caption", { class: "sr-only" }, "Products in this collection, in the shop's order"),
      h("thead", {}, h("tr", {}, h("th", { scope: "col" }, h("span", { class: "sr-only" }, "Photo")), h("th", { scope: "col" }, "Product"),
        h("th", { scope: "col" }, "Status"), h("th", { scope: "col" }, "Price"), h("th", { scope: "col" }, h("span", { class: "sr-only" }, "Order")))),
      h("tbody", {}, items.map(({ id, data: d }, i) => h("tr", {},
        h("td", { class: "t-img" }, d.images && d.images.length ? h("img", { src: thumb(d.images[0]), alt: "", width: "44", height: "44", loading: "lazy" }) : h("span", { class: "noimg" }, icon("image", 16))),
        h("td", {}, h("a", { href: "#/products/" + id, class: "t-name" }, d.name || id), h("span", { class: "t-id" }, id)),
        h("td", {}, d.published === false ? h("span", { class: "badge muted" }, "Draft") : h("span", { class: "badge green" }, "Active")),
        h("td", { class: "num" }, priceText(d)),
        h("td", { class: "t-move" }, moveBtn(items, i, -1), moveBtn(items, i, 1)))))));
    if (focusAfter) {
      // Keep the keyboard on the product that moved. At the top or the
      // bottom its button that way is switched off, so the other one takes it.
      const other = focusAfter.endsWith(":up") ? focusAfter.replace(/:up$/, ":down") : focusAfter.replace(/:down$/, ":up");
      let b = table.querySelector('[data-move="' + focusAfter + '"]');
      if (!b || b.disabled) b = table.querySelector('[data-move="' + other + '"]');
      if (b) b.focus();
      focusAfter = null;
    }
  }
  async function move(i, dir) {
    // One move at a time: a second click while the reorder is in flight would
    // send the list as it was before the reload and be refused, showing "not
    // saved" although the first move saved. The next click acts on the
    // reloaded list.
    if (moving) return;
    moving = true;
    const ids = data.items.map((x) => x.id);
    const [x] = ids.splice(i, 1);
    ids.splice(i + dir, 0, x);
    focusAfter = x + (dir < 0 ? ":up" : ":down");
    clear(msgs);
    try {
      await api("POST", "products/reorder", { body: { category: key, ids, expect_collection_rev: data.collection_rev } });
      toast("Order saved. The site rebuilt.");
    } catch (e) {
      msgs.append(banner({ tone: "critical", title: "The order was not saved.", items: [e.message] }));
    }
    try { await load(); } catch (e) { msgs.append(banner({ tone: "critical", title: e.message })); }
    moving = false;
  }
  load().catch((e) => msgs.append(banner({ tone: "critical", title: "The products did not load.", items: [e.message] })));
}

File: admin/bgsadmin/api/collections.py
"""Collections: the four categories and the unfiltered page, each edited as one.

A collection's words live in three documents: its name, breadcrumb name and
intro in copy.json, its circle in navigation.json (the entry that links to
collection.html?cat=<key>), and its homepage shelf in home.json. The screen
edits them as one view, and a save writes every document it touches in one
transaction, so the shop never shows a new name beside an old heading. The
view's rev covers only what the view holds: a save elsewhere in those
documents (the banner, the footer) never makes a collection stale. Which
products belong and in what order is the products' own data, changed with
products/reorder.
"""
import copy

from .. import lint
from .. import schema as schema_mod
from .. import validate
from ..errors import ApiError
from ..routes import Route
from ..schema.copy import COLLECTIONS
from ..schema.home import SHELVES
from ..service import changed, enforce, field_for, precondition, saved
from ..store.jsonstore import rev_of
from .meta import schemas

KEY = r"(?P<key>%s)" % "|".join(k for k, _ in COLLECTIONS)
NAMES = dict(COLLECTIONS)
SHELF_OF = {cat: sid for sid, _, cat, _ in SHELVES if cat}
DOCS = ("copy", "home", "navigation")


def href(key):
    return "collection.html" if key == "all" else "collection.html?cat=" + key


def nav_index(nav, key):
    """The category bar entry that opens this collection, or None. The
    unfiltered page has no circle of its own."""
    if key == "all":
        return None
    for i, c in enumerate(nav.get("categories") or []):
        if isinstance(c, dict) and c.get("href") == href(key):
            return i
    return None


def places(key, docs):
    """Each pointer in the view, and the document and pointer it is kept at."""
    out = {"/label": ("copy", "/categories/%s/label" % key),
           "/crumb": ("copy", "/categories/%s/crumb" % key),
           "/intro": ("copy", "/collection_intros/%s" % key)}
    i = nav_index(docs["navigation"], key)
    if i is not None:
        for part in ("label", "tint", "cutout", "image"):
            out["/circle/" + part] = ("navigation", "/categories/%d/%s" % (i, part))
    sid = SHELF_OF.get(key)
    if sid:
        out["/shelf/heading"] = ("home", "/sections/" + sid)
        out["/shelf/limit"] = ("home", "/shelves/%s/limit" % sid)
        out["/shelf/link_label"] = ("home", "/shelves/%s/link_label" % sid)
    return out


def view(key, docs):
    """What the collection screen edits, read out of the three documents. A
    collection with no circle or no shelf has null there, so the screen can
    say so instead of offering fields that would do nothing."""
    out = {"circle": None, "shelf": None}
    for vp, (doc, ptr) in places(key, docs).items():
        parts = vp.strip("/").split("/")
        holder = out
        if len(parts) == 2:
            if out[parts[0]] is None:
                out[parts[0]] = {}
            holder = out[parts[0]]
        v = validate.get(docs[doc], ptr)
        if v is not validate.MISSING:
            holder[parts[-1]] = copy.deepcopy(v)
    return out


def _walk(obj, ptr):
    """The container that holds ptr's last part, and that part. Missing
    objects on the way are created, so an older file gains the new keys."""
    parts = ptr.strip("/").split("/")
    for p in parts[:-1]:
        obj = obj[int(p)] if isinstance(obj, list) else obj.setdefault(p, {})
    return obj, (int(parts[-1]) if isinstance(obj, list) else parts[-1])


def apply(docs, where, cur, new):
    """Copies of the documents the change touches, with it written in. A
    pointer the view does not hold is refused, like any unknown field."""
    out = {}
    for vp in changed(cur, new):
        if vp not in where:
            raise ApiError(403, "not_editable", "That part of the collection cannot be changed here.", {"path": vp})
        doc, ptr = where[vp]
        target = out.setdefault(doc, copy.deepcopy(docs[doc]))
        holder, last = _walk(target, ptr)
        v = validate.get(new, vp)
        if v is validate.MISSING:
            if isinstance(holder, dict):
                holder.pop(last, None)
        else:
            holder[last] = copy.deepcopy(v)
    return out


def _docs(store):
    return {d: store.doc(d)[0] for d in DOCS}


def _counts(key, products):
    """Published and draft products in the collection, counted the way
    build.py counts them for a shelf's {n}."""
    rows = [d for d in products.values() if key == "all" or d.get("category") == key]
    live = sum(1 for d in rows if d.get("published", True))
    return live, len(rows) - live


def _shelf_shows(v, live):
    """How many cards the homepage shelf draws and what its link reads."""
    s = v.get("shelf")
    if not s:
        return None
    limit = s.get("limit")
    whole = isinstance(limit, int) and not isinstance(limit, bool)
    return {"cards": min(limit, live) if whole else live,
            "link": str(s.get("link_label") or "").replace("{n}", str(live))}


def _item(key, docs, products):
    v = view(key, docs)
    live, drafts = _counts(key, products)
    return {"key": key, "name": NAMES[key], "href": href(key), "rev": rev_of(v), "data": v,
            "published": live, "drafts": drafts, "shelf_shows": _shelf_shows(v, live)}


def fields(key, docs):
    """The schema field behind each pointer of the view, moved to that
    pointer, so the screen draws them the way it draws any document."""
    out = []
    for vp, (doc, ptr) in places(key, docs).items():
        f, _ = field_for(schemas()[doc]["fields"], ptr)
        if f:
            out.append(dict(f, path=vp))
    return out


def list_collections(req):
    store = req.app.store
    docs = _docs(store)
    products, crev = store.products()
    return {"collection_rev": crev, "items": [_item(k, docs, products) for k, _ in COLLECTIONS]}


def get_collection(req):
    store = req.app.store
    key = req.params["key"]
    docs = _docs(store)
    products, crev = store.products()
    return dict(_item(key, docs, products), fields=fields(key, docs), collection_rev=crev)


def put_collection(req):
    key = req.params["key"]
    body = req.body if isinstance(req.body, dict) else {}
    data = body.get("data")
    if not isinstance(data, dict):
        raise ApiError(400, "bad_request", "Send the collection as data.")
    store = req.app.store
    cfg = req.app.cfg
    sch = schemas()
    with store.transaction("edit collection %s" % key) as txn:
        docs = {d: txn.load(d) for d in DOCS}
        where = places(key, docs)
        cur = view(key, docs)
        precondition(req, rev_of(cur), cur)
        back = {v: k for k, v in where.items()}
        new_docs = apply(docs, where, cur, data)
        products, _ = store.products()
        ctx = {"icons": schema_mod.icons(cfg), "tints": schema_mod.tints(cfg), "products": products}
        errors = []
        for doc, new in new_docs.items():
            # Each document's own schema still decides what may change, so a
            # locked or read-only field (the circle's picture) refuses here
            # exactly as it does in the document's own editor.
            try:
                enforce(sch[doc]["fields"], docs[doc], new)
            except ApiError as e:
                if isinstance(e.details, dict) and (doc, e.details.get("path")) in back:
                    e.details["path"] = back[(doc, e.details["path"])]
                raise
            errs, _ = validate.document(doc, new, sch[doc]["fields"], ctx)
            errors += [dict(x, path=back.get((doc, x["path"]), x["path"]), document=doc) for x in errs]
        if errors:
            raise ApiError(422, "validation", "Some fields need attention.", errors)
        for doc, new in new_docs.items():
            txn.put(doc, new)
    # Site text that restates a store rule is checked as on any copy save.
    warnings = lint.after_save("copy", new_docs["copy"], store, sch) if "copy" in new_docs else []
    now = view(key, _docs(store))
    return saved(req.app, txn, key=key, rev=rev_of(now), data=now, warnings=warnings)


ROUTES = [
    Route("GET", r"collections", list_collections),
    Route("GET", r"collections/" + KEY, get_collection),
    Route("PUT", r"collections/" + KEY, put_collection, body="json"),
]
